feat: 웹소켓 with redis pub sub & simple broker

y2o2u2n opened the pull request:
#10 

- 작업 내용
  - 스케일 아웃시에도 메시지가 전달될 수 있도록 변경
  - redis pub/sub 을 사용해 메시지를 서버간 전달
- 참고 자료
  - https://dgempiuc.medium.com/spring-websocket-and-redis-pub-sub-a02af0dabddb

y2o2u2n committed:
chore: 패키지명 변경

y2o2u2n committed:
feat: redis pub sub & simple broker

y2o2u2n commented on ladder-domain/src/main/kotlin/kr/mashup/ladder/domain/room/domain/RoomChatSubscriber.kt:
```diff
@@ -0,0 +1,18 @@
+    override fun onMessage(message: Message, pattern: ByteArray?) {
+        val roomChat = JsonUtil.fromByteArray(message.body, RoomChat::class.java)
+        val event = RoomChatReceiveEvent(this, roomChat.roomId, roomChat.chat)
+        applicationEventPublisher.publishEvent(event)
```
api, domain 으로 모듈이 분리되어 있고
구독할 때는 redis 의존성이 있어야 하고
사용자에게 소켓으로 보내줄 때는 websocket 의존성이 있어야해서
분리하기 위해 이벤트를 사용했슴다

seungh0 commented on ladder-api/src/main/kotlin/kr/mashup/ladder/room/listener/RoomChatReceiveEventListener.kt:
```diff
@@ -0,0 +1,20 @@
+@Component
+class RoomChatReceiveEventListener(
+    private val simpMessagingTemplate: SimpMessagingTemplate,
+) : ApplicationListener<RoomChatReceiveEvent> {
```
이거 EventListener 사용하면,
ApplicationListener 구현 안해도 가능!

y2o2u2n replied:
좋다좋다
(멘션한 사람이 이모지 눌렀네 ㅋㅋ...)

seungh0 commented on ladder-domain/src/main/kotlin/kr/mashup/ladder/domain/room/domain/RoomChatReceiveEvent.kt:
```diff
@@ -0,0 +1,9 @@
+    source: Any,
+    val roomId: Long,
+    val chat: String
+) : ApplicationEvent(source)
```
요것두 ApplicationEvent 필요 없어져

y2o2u2n replied:
좋다좋다 의존성 제거 굿...

seungh0 commented on ladder-domain/src/main/kotlin/kr/mashup/ladder/domain/config/redis/EmbeddedRedisServerConfig.kt:
```diff
@@ -0,0 +1,32 @@
+) {
+    companion object {
+        private var redisServer: RedisServer? = null
+    }
```
요거 companion object 가 맨 아래에 있는게 컨벤션인거 같은데.. 아닌가 ( 잘 모름)

자바  `private static final ` 맨 위에 두다가 맨 아래에 두니 어색했는데 ㅋㅋㅋ

seungh0 replied:
private var lateinit redisServer: RedisServer 

로 nullable 하게 해도 좋을거 같아!

(이게 안되서 이렇게 헀나??)

y2o2u2n replied:
음... 이게 테스트1, 테스트2, 테스트3 이 있다고 치면

테스트1 이 끝나고 서버를 중지시켜야 되는데
테스트 내부적으로 캐싱해서 쓰는게 있어서
테스트2가 먼저 시작되서 실패하는 경우가 있더라고

이 부분은 local, test 별개로 가져가서
local 은 지금처럼 쓰고
test 는 테스트 전마다 실행하도록 `@BeforeEach` 에서 재실행하는걸로 변경해야겠다.

그래서 지금 상황은 테스트 실행간 계속 실행되도록 켜놓은거긴해서 아주 꼼꼼히 잘 봐줬슴다...

이 부분은 내일 만나서 수정하는걸루... ㅎㅎㅎ

seungh0 reviewed (approved):
잘 배워갑니다 ㅎㅎㅎ

갑자기 궁금한게

 소켓 서버로 커넥션이 연결되 있는 상황에서 해당 서버가 배포 or 장애 등으로 down 되면
거기에 물려있던 커넥션들은 어떻게 처리돼?

클라에서 다른 서버로 커넥션 맺도록, Retry 처리하는 방식으로 사용하려나..?

y2o2u2n commented:
> 잘 배워갑니다 ㅎㅎㅎ
> 
> 갑자기 궁금한게
> 
> 소켓 서버로 커넥션이 연결되 있는 상황에서 해당 서버가 배포 or 장애 등으로 down 되면 거기에 물려있던 커넥션들은 어떻게 처리돼?
> 
> 클라에서 다른 서버로 커넥션 맺도록, Retry 처리하는 방식으로 사용하려나..?

웹소켓 커넥션은 각 서버들이 들고 있기 때문에 스프링이 다운되면 커넥션이 다 끊길 것 같고
그런 상황이라면 다른 서버로 다시 연결을 맺도록 해야 될 것 같아
(커넥션이 정상인지는 heartbeat 로 하는데 지금은 디폴트 25초로 되어있넹... (더 짧게 잡아야할 수도))

seungh0 commented:
아하아하 땡큐

y2o2u2n committed:
refactor: 코드 리뷰

- url mapping 전체 경로 명시
- 샘플 코드 deprecated 처리
- 테이블명 변경 - prefix 제거
- dto 내 request, response 분리
- 임베디드 레디스 버전 변경
- slf4j 의존성 충돌 문제로 한 버전 낮은것 사용
- 스프링 이벤트 상속 대신 애너테이션 사용
